### PR TITLE
Upgrade to LLVM 14

### PR DESCRIPTION
**Pull request by ptersilie**
This PR upgrades our LLVM fork to 14.x.

Note that in order to do that we first had to manually push a new branch `ykjit/14.x` which copies branch `llvm-project/release/14.x`. This PR then cherry-picks all our changes from `ykjit/13.x` onto this branch, fixing any issues that arose from the upgrade in 7ec87521ffe74a991a85c576bfc966c32883f1c3.

Also note, that cherry-picking doesn't like if a range of commits contains merge commits. It's possible to work around this which creates empty merge commits. I had the choice to skip them, but decided to keep them in so it's still possible from the history to work out which PR a change came from. I'm not strongly opposed to removing them if others prefer that.

Companion PR: https://github.com/ykjit/yk/pull/557

### CODE REVIEW

**Comment by ltratt** [comment]
To the extent I can tell this looks fine to me!

File: llvm/lib/LTO/LTOBackend.cpp [hunk 1]
```diff
@@ -79,11 +79,6 @@ namespace llvm {
 extern cl::opt<bool> NoPGOWarnMismatch;
 }
 
-static cl::opt<bool>
```
Comment (vext01) [inline llvm/lib/LTO/LTOBackend.cpp]: How comes this disappeared? Was it moved elsewhere in another commit perhaps?

Comment (ptersilie) [reply llvm/lib/LTO/LTOBackend.cpp]: LLVM 14 didn't like that we had two of them. The other one that I kept lives here: https://github.com/ykjit/ykllvm/blob/ykjit/13.x/llvm/lib/CodeGen/TargetPassConfig.cpp#L247

Comment (ptersilie) [reply llvm/lib/LTO/LTOBackend.cpp]: A possible explanation is that those two weren't in the same namespace previously which is why we needed both, but LLVM 14 changed the namespaces so now they were and thus LLVM wasn't happy.

Comment (vext01) [reply llvm/lib/LTO/LTOBackend.cpp]: If there is duplication, then deleting one is certainly correct!

**Comment by vext01** [comment]
I think this one is ready to go?

**Comment by ltratt** [comment]
Fine with me!
